server/src/main/java/io/crate/planner/operators/Union.java

```diff
@@ -240,6 +240,12 @@ public LogicalPlan pruneOutputsExcept(SequencedCollection<Symbol> outputsToKeep)
         if (newLhs == lhs && newRhs == rhs) {
             return this;
         }
+        if (newLhs.outputs().size() != newRhs.outputs().size()) {
+            // UNION left/right side plans must have outputs of the same size.
+            // Column prunning can lead to a situation when one side stays intact and another gets modified.
+            // We should not use prunned plans in this case to keep UNION operator valid.
+            return this;
+        }
```

Comment:
I don't think that this is correct, also I wonder how this could happen as the `outputsToKeep` are dictated above, so no relation should emit more outputs.

The problem with the query used in the issue is more related that the `HashAggregate` of the `count(*)` wants to prune *all* outputs, thus the `outputsToKeep` is empty and this seems not to be honored by the left and right relation tree. Looks like there is an `Eval` operator above the tablefunction with an output not being pruned.

Reply:
>Looks like there is an Eval operator above the tablefunction with an output not being pruned.

Because TableFunction with empty `toCollect` returns same instance. when prunned with empty list.

>and Rename behaving differently than it's stated in javadocs

That was about scenario with View, ie there right side of the UNION is Rename.


>update all Operators to return same instance if outputsToKeep is am empty List

I will update all operators to do that ^ but I remember I had quite some failures, will double check what's happening in this case.

Reply:
>That was about scenario with View, ie there right side of the UNION is Rename.

Just checked, it's almost, same , Union ->Rename-> Eval -> TableFunction. 

Will fix `TableFunction`

Reply:
>thus the outputsToKeep is empty and this seems not to be honored by the left and right relation tree. Looks like there is an Eval operator above the tablefunction with an output not being pruned.

Thanks, I removed this and switched to another approach.

I checked `Eval` implementation, it keeps the output (even with `outputsToKeep` being empty) because source (TableFunction) stays intact. And TableFunction does the same.

Both Eval and TableFunction explicitly follow the contract from the javadoc
> If there are no outputs to prune and if the source also didn't change, `this` must be returned.

Problem with UNION is that right side (Eval->TableFunction) follows the contract and doesn't mutate the plan but Collect apparently doesn't (ie mutates plan) which leads to left/right sides of the UNION being relations with different size outputs.

I pushed a fix to make Collect to follow the contract.